frontend/packages/topology/src/components/graph-view/Topology.tsx

```diff
@@ -186,14 +186,24 @@ const Topology: React.FC<TopologyProps &
     });
     setVisualization(newVisualization);
     return newVisualization;
-  };
+  }, [
+    namespace,
+    onGraphModelChange,
+    onSelect,
+    setTopologyLayoutData,
+    setVisualization,
+    topologyLayoutDataJson,
+  ]);
 
   const visualizationRef = React.useRef<Visualization>();
-  if (!visualizationRef.current) {
-    visualizationRef.current = createVisualization();
-  }
   const visualization = visualizationRef.current;
 
+  React.useEffect(() => {
+    if (!visualizationRef.current) {
+      visualizationRef.current = createVisualization();
+    }
+  }, [createVisualization]);
+
```

Comment:
Updating a `ref` in the render flow is ok. This doesn't require a change for it to be done inside an effect.

Reply:
there is a setState call inside `createVisualization`, that's why moved the ref inside an effect.

Reply:
I didn't realize that `createVisualization` was executing a callback which could set state. In that case what about:
```
  if (!visualizationRef.current) {
    visualizationRef.current = createVisualization();
  }
  const visualization = visualizationRef.current;
  React.useEffect(() => {
    if (visualization) {
      setVisualization(visualization);
    }
  }, [visualization]);
```

Reply:
Make sure to also fix-up `TopologyListView` which does something similar.

Reply:
@christianvogt fixed